From b2353bc99d31dd41e3f4a03b993bc13b2e2f4802 Mon Sep 17 00:00:00 2001
From: Michael Peyton Jones <me@michaelpj.com>
Date: Sat, 30 Oct 2021 18:08:23 +0100
Subject: [PATCH] Aeson 2 compatibility

We get compatibility with both <2 and >=2 by using only functions that
appear in both, which don't make assumptions about the structure of
objects.

Fixes #356
---
 lsp-types/src/Language/LSP/Types/Common.hs  |  3 +-
 lsp-types/src/Language/LSP/Types/Message.hs |  6 +--
 lsp-types/src/Language/LSP/Types/Parsing.hs | 56 +++++++++++----------
 lsp/example/Reactor.hs                      |  5 +-
 4 files changed, 35 insertions(+), 35 deletions(-)

diff --git a/lsp/example/Reactor.hs b/lsp/example/Reactor.hs
index cbe7d796..a4c3a817 100644
--- a/lsp/example/Reactor.hs
+++ b/lsp/example/Reactor.hs
@@ -30,7 +30,6 @@ import           Control.Monad
 import           Control.Monad.IO.Class
 import           Control.Monad.STM
 import qualified Data.Aeson                            as J
-import qualified Data.HashMap.Strict                   as H
 import qualified Data.Text                             as T
 import           GHC.Generics (Generic)
 import           Language.LSP.Server
@@ -263,8 +262,8 @@ handle = mconcat
               cmd = "lsp-hello-command"
               -- need 'file' and 'start_pos'
               args = J.List
-                      [ J.Object $ H.fromList [("file",     J.Object $ H.fromList [("textDocument",J.toJSON doc)])]
-                      , J.Object $ H.fromList [("start_pos",J.Object $ H.fromList [("position",    J.toJSON start)])]
+                      [ J.object [("file",     J.object [("textDocument",J.toJSON doc)])]
+                      , J.object [("start_pos",J.object [("position",    J.toJSON start)])]
                       ]
               cmdparams = Just args
           makeCommand (J.Diagnostic _r _s _c _source _m _t _l) = []
From c974bb0eaa204791c5abb7d051fe5b92ee018d80 Mon Sep 17 00:00:00 2001
From: Kobayashi <contact@zelinf.net>
Date: Tue, 4 Jan 2022 22:20:20 +0800
Subject: [PATCH] fix textDocument/selectionRange's method (#388)

* fix selection range

* remove Arbitrary instance for Value when aeson provides it
---
 lsp-types/src/Language/LSP/Types/Method.hs | 2 +-
 lsp/test/JsonSpec.hs                       | 3 +++
 2 files changed, 4 insertions(+), 1 deletion(-)

diff --git a/lsp/test/JsonSpec.hs b/lsp/test/JsonSpec.hs
index 7d77e940..ab95dbf0 100644
--- a/lsp/test/JsonSpec.hs
+++ b/lsp/test/JsonSpec.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP                  #-}
 {-# LANGUAGE UndecidableInstances #-}
 {-# LANGUAGE FlexibleContexts     #-}
 {-# LANGUAGE TypeInType           #-}
@@ -159,6 +160,7 @@ smallList = resize 3 . listOf
 instance (Arbitrary a) => Arbitrary (List a) where
   arbitrary = List <$> arbitrary
 
+#if !MIN_VERSION_aeson(2,0,3)
 instance Arbitrary J.Value where
   arbitrary = oneof
     [ J.String <$> arbitrary
@@ -166,6 +168,7 @@ instance Arbitrary J.Value where
     , J.Bool <$> arbitrary
     , pure J.Null
     ]
+#endif
 
 -- ---------------------------------------------------------------------
 
